### ISSUE TITLE
[Feature] Multiple Users with Permissions

### ISSUE THREAD

**Issue by richardfoltin** [stashapp/stash#2337]
**Is your feature request related to a problem? Please describe.**
I would like to have my videos at a safe place. So my idea is that I would create 2 accounts for myself; an admin and a user. User account credentials are saved in the browser thus it is easy to access the videos. Admin account is only used if organinizing is needed.

**Describe the solution you'd like**
I would like to have the possiblity to have multiple accouns and to configure the permissions that an account can have. Especially in case of important features like Deleting videos.

**Comment by QxxxGit** [comment]
I think multiple users with a permissions system could be pretty cool for Stash.

You could possibly see a wide array of uses if it's within the scope of the devs vision for Stash. Amateur/professionals could host their own install, allowing subscribers under a user role to see specific scenes/images/galleries. You would probably see some tube sites start using Stash to host content. Even on a local network, user permissions would be useful. My initial thoughts:

* Multiple users would necessitate an additional table for storing user data (name, pass, etc.). This table would evolve for feature requests (email validation and what have you).
* Permissions would range from viewing/creating/editing records (scenes, performers, etc.), managing users, and so on.
* Permissions could be set for individual media, collections/galleries, and for user roles. 

Of course something of this depth would be much further down the road; but I think it's worth getting ideas written out in advance.

**Comment by kermieisinthehouse** [comment]
> You would probably see some tube sites start using Stash to host content. 

I think that, partially, using sqlite, and primarily, reading directly from block storage means that this will never scale past 25 users. It would be a huge rewrite to make stash "scalable"

**Comment by RRocks01** [comment]
I came here to see if this was possible, I just wanted to create permissions for my wife, and ideally let her have her own "like" count, and more importantly keep mine private and library settings.  Would be nice to restrict content too.

**Comment by WithoutPants** [comment]
This is something I'd like to add at some point. As others have pointed out, it will require a substantial amount of effort.

**Comment by sleetx** [comment]
A multi-user environment would be great. We'll need a place in the settings where accounts can be managed. Plus a minimum of two assignable roles: admin and user. Admin is what we have now. Here are some requirements for a limited "user" role that is mostly read-only:

Access restrictions:
- No access to the "Settings" page (this includes browsing to the /settings URL manually)
- No access to the "Edit" tab on any page. (The "rating" option is here on some of them, and we probably want to let users rate things on their own. So ratings should be moved to another tab.)
- No access to the Tagger view or "Auto Tag" buttons
- No access to the "Operations" button on any page (allowing rescan, item deletion, etc)
- No access to the "Organized" button
- No access to the "Submit to Stash-Box" button
- The "File info" tab I am on the fence about... do users need to know those details?
- Also not sure if we want them creating markers
- The Help menu would be role-dependent; the "user" version should not reference the admin panel

Things that should be unique to each user account (rather than globally set):
- Filters - let them manage their own
- Homepage - let them set their own homepage
- Ratings
- O-counter

**Comment by HoLLy-HaCKeR** [comment]
A single read-only user in addition to the admin user may also serve a lot of usecases, and would require significantly less effort to implement.

**Comment by skier233** [comment]
+1

**Comment by byertsy** [comment]
+1

**Comment by DogmaDragon** [comment]
@byertsy leave 👍 or 👎 on the first post instead of `+1` and other upvote comments.

**Comment by Mlitz** [comment]
Any Update on this?

**Comment by rohitghali** [comment]
Any plans to implement this? Doesn't have to be scalable initially. Few users for local use would be sufficient. Kind of like what jellyfin implemented.

**Comment by WithoutPants** [comment]
> Any plans to implement this? Doesn't have to be scalable initially. Few users for local use would be sufficient. Kind of like what jellyfin implemented.

This is in the roadmap (which is not currently published) and I've added it to the backlog to signify this. I have a WIP branch if someone wants to pick up the work, but I've got a number of higher priority things to work on before I can properly implement this.

- https://github.com/stashapp/stash/pull/6503

**Comment by Blabss0** [comment]
Would be amazing to have this feature

**Comment by einfachJens** [comment]
+1

**Comment by upamanyudas** [comment]
+1

**Comment by TheZitroX** [comment]
+1

**Comment by TheZitroX** [comment]
3y no Multi User?

**Comment by BillyOutlast** [comment]
> 3y no Multi User?

Because of you!!! >:#

But fr fr, I know, I'm so sad, I wish the Plex integration was better:(

**Comment by TheZitroX** [comment]
Marked as `spam`? What? Dont know what `+1` means?

**Comment by DogmaDragon** [comment]
> Marked as `spam`? What? Dont know what `+1` means?

@TheZitroX Changed to off-topic as there isn't a better category to place it under. Please use 👍 instead of unhelpful comments to show support for a feature request.

**Comment by PilaScat** [comment]
Any news? at lest a read only user, initially would be perfect

**Comment by GalacticCheat** [comment]
Maista sucks cock
![Image](https://github.com/user-attachments/assets/ed12369e-cb07-434a-b49f-fbc470272035)
https://i.ibb.co/ds49xcXt/image-2-copie.jpg

**Comment by discourse-stashapp** [comment]
This issue has been mentioned on **Stash**. There might be relevant details there:

https://discourse.stashapp.cc/t/multiple-user-accounts/4829/2

**Comment by dillardblom** [comment]
> A single read-only user in addition to the admin user may also serve a lot of usecases, and would require significantly less effort to implement.

I do agree with this view. I do realise making a full user-system with management, ACL is gonna be a lot of work. But a 'in-between' solution of two users, where one is the normal (current) user and one can only view, not edit not delete would be a quick(?) to-implement solution, that allows users to invite other people to watch your collection, without the risk of someone accidentally editting, updating or deleting content.

**Comment by trulow** [comment]
Has there been any updates or progress to this? This would be very helpful as I would like to share read only access to a friend.

**Comment by tylercranston** [comment]
If users cannot be easily implemented, it would be nice to have a flag in the config where the application goes into a read only mode. This should prevent users from adding, editing,
or deleting content but should allow them to view any content and also should incement view counters and last play timestamp as an example.

**Comment by Yellowfin34** [comment]
https://github.com/Yellowfin34/stash-multiuser

<div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h2 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; border-bottom: 1px solid rgba(61, 68, 77, 0.7); padding-bottom: 0.3em;">Project goals</h2><a id="user-content-project-goals" class="anchor" aria-label="Permalink: Project goals" href="https://github.com/Yellowfin34/stash-multiuser#project-goals" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 19.0938px; left: -28px; transform: translateY(calc(-50% - 0.3rem));"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;"><strong style="box-sizing: border-box; font-weight: 600;">Shared global catalog:</strong><span> </span>scenes, performers, studios, tags, galleries/albums, files, folders, fingerprints, scrapers, and scene markers remain canonical/global records.</li><li style="box-sizing: border-box; margin-top: 0.25em;"><strong style="box-sizing: border-box; font-weight: 600;">Private per-user state:</strong><span> </span>users get separate watch history, resume position, ratings, likes, favorites, O-counts, hidden state, and future playlist/preferences support.</li><li style="box-sizing: border-box; margin-top: 0.25em;"><strong style="box-sizing: border-box; font-weight: 600;">Role-based access control:</strong><span> </span>owners/admins manage the system; curators can edit shared metadata; viewers can watch and keep personal state but cannot add or mutate shared library records.</li><li style="box-sizing: border-box; margin-top: 0.25em;"><strong style="box-sizing: border-box; font-weight: 600;">Admin visibility:</strong><span> </span>owners/admins can see aggregated user activity/statistics without exposing unrestricted raw history to regular users.</li><li style="box-sizing: border-box; margin-top: 0.25em;"><strong style="box-sizing: border-box; font-weight: 600;">Safe production operation:</strong><span> </span>repeatable migration, Docker deployment, backups, restore drills, and reverse-proxy guidance are documented in-repo.</li></ul><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h2 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; border-bottom: 1px solid rgba(61, 68, 77, 0.7); padding-bottom: 0.3em;">Current role model</h2><a id="user-content-current-role-model" class="anchor" aria-label="Permalink: Current role model" href="https://github.com/Yellowfin34/stash-multiuser#current-role-model" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 19.0938px; left: -28px; transform: translateY(calc(-50% - 0.3rem));"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><markdown-accessiblity-table data-catalyst="" style="box-sizing: border-box; display: block; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">
Role | Media access | Personal activity | Add/edit shared metadata | Settings/system tasks | User admin/activity stats
-- | -- | -- | -- | -- | --
owner | yes | yes | yes | yes | yes
admin | yes | yes | yes | yes | yes
curator | yes | yes | yes | no | no
viewer | yes | yes | no | no | no

</markdown-accessiblity-table><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Shared metadata includes scenes, performers, tags, studios, groups/albums, galleries, gallery chapters, scene markers, and related global library records.</p><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h2 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; border-bottom: 1px solid rgba(61, 68, 77, 0.7); padding-bottom: 0.3em;">Major changes in this fork</h2><a id="user-content-major-changes-in-this-fork" class="anchor" aria-label="Permalink: Major changes in this fork" href="https://github.com/Yellowfin34/stash-multiuser#major-changes-in-this-fork" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 19.0938px; left: -28px; transform: translateY(calc(-50% - 0.3rem));"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25;">Authentication and users</h3><a id="user-content-authentication-and-users" class="anchor" aria-label="Permalink: Authentication and users" href="https://github.com/Yellowfin34/stash-multiuser#authentication-and-users" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 12.5px; left: -28px; transform: translateY(-50%);"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;">Added<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">users</code><span> </span>and<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">user_sessions</code><span> </span>tables via schema migration<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">86_multi_user</code>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added bcrypt password hashing and reusable password validation.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Enforced password policy: at least<span> </span><strong style="box-sizing: border-box; font-weight: 600;">5 characters</strong><span> </span>and at least<span> </span><strong style="box-sizing: border-box; font-weight: 600;">1 special character</strong>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added GraphQL auth/account operations:<ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px; padding-left: 2em;"><li style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">currentUser</code></li><li style="box-sizing: border-box; margin-top: 0.25em;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">login</code></li><li style="box-sizing: border-box; margin-top: 0.25em;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">logout</code></li><li style="box-sizing: border-box; margin-top: 0.25em;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">users</code></li><li style="box-sizing: border-box; margin-top: 0.25em;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">userCreate</code></li><li style="box-sizing: border-box; margin-top: 0.25em;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">userUpdate</code></li><li style="box-sizing: border-box; margin-top: 0.25em;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">userResetPassword</code></li></ul></li><li style="box-sizing: border-box; margin-top: 0.25em;">Aligned existing form-cookie login/logout with DB-backed multi-user sessions.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added current-user context wiring through the existing API/auth middleware.</li></ul><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25;">Per-user scene activity</h3><a id="user-content-per-user-scene-activity" class="anchor" aria-label="Permalink: Per-user scene activity" href="https://github.com/Yellowfin34/stash-multiuser#per-user-scene-activity" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 12.5px; left: -28px; transform: translateY(-50%);"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;">Added<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">user_scene_activity</code><span> </span>support for:<ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px; padding-left: 2em;"><li style="box-sizing: border-box;">last watched timestamp</li><li style="box-sizing: border-box; margin-top: 0.25em;">resume seconds</li><li style="box-sizing: border-box; margin-top: 0.25em;">watch count</li><li style="box-sizing: border-box; margin-top: 0.25em;">watch seconds</li><li style="box-sizing: border-box; margin-top: 0.25em;">O-count</li><li style="box-sizing: border-box; margin-top: 0.25em;">liked state</li><li style="box-sizing: border-box; margin-top: 0.25em;">favorite state</li><li style="box-sizing: border-box; margin-top: 0.25em;">hidden state</li><li style="box-sizing: border-box; margin-top: 0.25em;">personal rating</li></ul></li><li style="box-sizing: border-box; margin-top: 0.25em;">Added GraphQL scene activity API:<ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px; padding-left: 2em;"><li style="box-sizing: border-box;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">sceneActivity</code></li><li style="box-sizing: border-box; margin-top: 0.25em;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">sceneActivityUpdate</code></li></ul></li><li style="box-sizing: border-box; margin-top: 0.25em;">Updated scene model resolvers so user-specific scene state comes from the current user's row where applicable.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added tests around store behavior and resolver current-user isolation.</li></ul><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25;">Admin “God mode” activity statistics</h3><a id="user-content-admin-god-mode-activity-statistics" class="anchor" aria-label="Permalink: Admin “God mode” activity statistics" href="https://github.com/Yellowfin34/stash-multiuser#admin-god-mode-activity-statistics" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 12.5px; left: -28px; transform: translateY(-50%);"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Owners/admins can view aggregated user statistics in the<span> </span><strong style="box-sizing: border-box; font-weight: 600;">Users and permissions</strong><span> </span>settings panel, including:</p><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;">total plays</li><li style="box-sizing: border-box; margin-top: 0.25em;">total watch seconds</li><li style="box-sizing: border-box; margin-top: 0.25em;">total O-count</li><li style="box-sizing: border-box; margin-top: 0.25em;">liked/rated/favorite/hidden counts</li><li style="box-sizing: border-box; margin-top: 0.25em;">last watched timestamp</li><li style="box-sizing: border-box; margin-top: 0.25em;">recent per-scene activity summaries</li></ul><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Access is protected by the<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">manage_users</code>/administrator guard so viewer/curator accounts cannot query or view this data.</p><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25;">RBAC and API hardening</h3><a id="user-content-rbac-and-api-hardening" class="anchor" aria-label="Permalink: RBAC and API hardening" href="https://github.com/Yellowfin34/stash-multiuser#rbac-and-api-hardening" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 12.5px; left: -28px; transform: translateY(-50%);"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The fork adds backend authorization guards for shared metadata and system-sensitive operations. Viewers are blocked server-side from creating/updating/deleting shared library records even if they manually call GraphQL.</p><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Guarded areas include:</p><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;">scene create/update/delete and bulk scene metadata operations</li><li style="box-sizing: border-box; margin-top: 0.25em;">performer create/update/delete</li><li style="box-sizing: border-box; margin-top: 0.25em;">tag create/update/delete/merge</li><li style="box-sizing: border-box; margin-top: 0.25em;">studio create/update/delete</li><li style="box-sizing: border-box; margin-top: 0.25em;">group/album create/update/delete</li><li style="box-sizing: border-box; margin-top: 0.25em;">gallery and gallery chapter create/update/delete</li><li style="box-sizing: border-box; margin-top: 0.25em;">scene markers</li><li style="box-sizing: border-box; margin-top: 0.25em;">metadata scan/generate/identify/auto-tag/import/export jobs</li><li style="box-sizing: border-box; margin-top: 0.25em;">database clean/backup/anonymize operations</li><li style="box-sizing: border-box; margin-top: 0.25em;">package/plugin install/update/uninstall/reload operations</li><li style="box-sizing: border-box; margin-top: 0.25em;">scraper reload</li><li style="box-sizing: border-box; margin-top: 0.25em;">job queue/subscription/stop operations</li><li style="box-sizing: border-box; margin-top: 0.25em;">settings/configuration mutations</li><li style="box-sizing: border-box; margin-top: 0.25em;">API key generation/clear operations</li></ul><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25;">Frontend changes</h3><a id="user-content-frontend-changes" class="anchor" aria-label="Permalink: Frontend changes" href="https://github.com/Yellowfin34/stash-multiuser#frontend-changes" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 12.5px; left: -28px; transform: translateY(-50%);"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;">Added current-user role helpers in<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">ui/v2.5/src/core/multiuser.ts</code>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added multi-user user/admin UI in Settings →<span> </span><strong style="box-sizing: border-box; font-weight: 600;">Users and permissions</strong>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added admin activity display in the same settings panel.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Hid Settings/admin navigation for non-admins.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Blocked direct<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">/settings</code><span> </span>access for non-admins.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Hid the<span> </span><strong style="box-sizing: border-box; font-weight: 600;">New</strong><span> </span>button and disabled the<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">n</code><span> </span>hotkey for users that cannot edit shared metadata.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added frontend role/helper tests.</li></ul><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25;">Migration tooling</h3><a id="user-content-migration-tooling" class="anchor" aria-label="Permalink: Migration tooling" href="https://github.com/Yellowfin34/stash-multiuser#migration-tooling" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 12.5px; left: -28px; transform: translateY(-50%);"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;">Added repeatable CLI migrator:<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">cmd/stash-multiuser-migrate</code>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added migration docs:<span> </span><a href="https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docs/MULTIUSER_DATABASE_MIGRATION.md" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; text-underline-offset: 0.2rem;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">docs/MULTIUSER_DATABASE_MIGRATION.md</code></a>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Migration keeps the existing Stash catalog global and seeds the first owner account.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Existing single-user activity is imported into the first owner account where possible.</li><li style="box-sizing: border-box; margin-top: 0.25em;">The migrator is intended to run against a copied database, never against the source server live DB.</li></ul><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 tabindex="-1" class="heading-element" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25;">Production/deployment tooling</h3><a id="user-content-productiondeployment-tooling" class="anchor" aria-label="Permalink: Production/deployment tooling" href="https://github.com/Yellowfin34/stash-multiuser#productiondeployment-tooling" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; float: left; padding-right: 4px; margin: auto; line-height: 1; text-underline-offset: 0.2rem; border-radius: 6px; opacity: 0; justify-content: center; align-items: center; width: 28px; height: 28px; display: flex; position: absolute; top: 12.5px; left: -28px; transform: translateY(-50%);"><svg data-component="Octicon" class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></div><ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; padding-left: 2em; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;">Added Docker deployment runbook:<span> </span><a href="https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docs/MULTIUSER_DOCKER_DEPLOYMENT.md" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; text-underline-offset: 0.2rem;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">docs/MULTIUSER_DOCKER_DEPLOYMENT.md</code></a>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added production readiness checklist:<span> </span><a href="https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docs/PRODUCTION_READINESS.md" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; text-underline-offset: 0.2rem;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">docs/PRODUCTION_READINESS.md</code></a>.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added backup script and restore-drill documentation.</li><li style="box-sizing: border-box; margin-top: 0.25em;">Added NGINX reverse-proxy template for<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">self-service.fun</code>:<ul dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px; padding-left: 2em;"><li style="box-sizing: border-box;"><a href="https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docker/production/nginx-self-service.fun.conf" style="box-sizing: border-box; background-color: rgba(0, 0, 0, 0); color: rgb(68, 147, 248); text-decoration: underline; text-underline-offset: 0.2rem;"><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">docker/production/nginx-self-service.fun.conf</code></a></li></ul></li><li style="box-sizing: border-box; margin-top: 0.25em;">The app container is expected to run HTTP internally on port<span> </span><code style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 13.6px; tab-size: 4; white-space: break-spaces; background-color: rgba(101, 108, 118, 0.2); border-radius: 6px; margin: 0px; padding: 0.2em 0.4em;">9999</code>; TLS/public exposure should be handled by a dedicated NGINX reverse proxy.</li></ul><div class="markdown-heading" dir="auto" style="box-sizing: border-box; position: relative; color: rgb(240, 246, 252); font-family: &quot;Mona Sans VF&quot;, -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"></div><br class="Apple-interchange-newline">Project goals
Shared global catalog: scenes, performers, studios, tags, galleries/albums, files, folders, fingerprints, scrapers, and scene markers remain canonical/global records.
Private per-user state: users get separate watch history, resume position, ratings, likes, favorites, O-counts, hidden state, and future playlist/preferences support.
Role-based access control: owners/admins manage the system; curators can edit shared metadata; viewers can watch and keep personal state but cannot add or mutate shared library records.
Admin visibility: owners/admins can see aggregated user activity/statistics without exposing unrestricted raw history to regular users.
Safe production operation: repeatable migration, Docker deployment, backups, restore drills, and reverse-proxy guidance are documented in-repo.
Current role model
Role	Media access	Personal activity	Add/edit shared metadata	Settings/system tasks	User admin/activity stats
owner	yes	yes	yes	yes	yes
admin	yes	yes	yes	yes	yes
curator	yes	yes	yes	no	no
viewer	yes	yes	no	no	no
Shared metadata includes scenes, performers, tags, studios, groups/albums, galleries, gallery chapters, scene markers, and related global library records.

Major changes in this fork
Authentication and users
Added users and user_sessions tables via schema migration 86_multi_user.
Added bcrypt password hashing and reusable password validation.
Enforced password policy: at least 5 characters and at least 1 special character.
Added GraphQL auth/account operations:
currentUser
login
logout
users
userCreate
userUpdate
userResetPassword
Aligned existing form-cookie login/logout with DB-backed multi-user sessions.
Added current-user context wiring through the existing API/auth middleware.
Per-user scene activity
Added user_scene_activity support for:
last watched timestamp
resume seconds
watch count
watch seconds
O-count
liked state
favorite state
hidden state
personal rating
Added GraphQL scene activity API:
sceneActivity
sceneActivityUpdate
Updated scene model resolvers so user-specific scene state comes from the current user's row where applicable.
Added tests around store behavior and resolver current-user isolation.
Admin “God mode” activity statistics
Owners/admins can view aggregated user statistics in the Users and permissions settings panel, including:

total plays
total watch seconds
total O-count
liked/rated/favorite/hidden counts
last watched timestamp
recent per-scene activity summaries
Access is protected by the manage_users/administrator guard so viewer/curator accounts cannot query or view this data.

RBAC and API hardening
The fork adds backend authorization guards for shared metadata and system-sensitive operations. Viewers are blocked server-side from creating/updating/deleting shared library records even if they manually call GraphQL.

Guarded areas include:

scene create/update/delete and bulk scene metadata operations
performer create/update/delete
tag create/update/delete/merge
studio create/update/delete
group/album create/update/delete
gallery and gallery chapter create/update/delete
scene markers
metadata scan/generate/identify/auto-tag/import/export jobs
database clean/backup/anonymize operations
package/plugin install/update/uninstall/reload operations
scraper reload
job queue/subscription/stop operations
settings/configuration mutations
API key generation/clear operations
Frontend changes
Added current-user role helpers in ui/v2.5/src/core/multiuser.ts.
Added multi-user user/admin UI in Settings → Users and permissions.
Added admin activity display in the same settings panel.
Hid Settings/admin navigation for non-admins.
Blocked direct /settings access for non-admins.
Hid the New button and disabled the n hotkey for users that cannot edit shared metadata.
Added frontend role/helper tests.
Migration tooling
Added repeatable CLI migrator: cmd/stash-multiuser-migrate.
Added migration docs: [docs/MULTIUSER_DATABASE_MIGRATION.md](https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docs/MULTIUSER_DATABASE_MIGRATION.md).
Migration keeps the existing Stash catalog global and seeds the first owner account.
Existing single-user activity is imported into the first owner account where possible.
The migrator is intended to run against a copied database, never against the source server live DB.
Production/deployment tooling
Added Docker deployment runbook: [docs/MULTIUSER_DOCKER_DEPLOYMENT.md](https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docs/MULTIUSER_DOCKER_DEPLOYMENT.md).
Added production readiness checklist: [docs/PRODUCTION_READINESS.md](https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docs/PRODUCTION_READINESS.md).
Added backup script and restore-drill documentation.
Added NGINX reverse-proxy template for self-service.fun:
[docker/production/nginx-self-service.fun.conf](https://github.com/Yellowfin34/stash-multiuser/blob/multiuser/main/docker/production/nginx-self-service.fun.conf)
The app container is expected to run HTTP internally on port 9999; TLS/public exposure should be handled by a dedicated NGINX reverse proxy.

**Comment by Experiment-696** [comment]
I essentially have this already implemented multiple users multiple roles with permissions (Admin Moderator User)
Users:
Can Access Scenes 
Have their own O-Counter and Rating(as well as global average rating + # of rates)
Have their own Filters
Can customize their theme (reverts to global theme on logout/guest view)
everything else forbidden
Moderators: all of user permissions plus
can edit scene and performer metadata of new scenes with no metadata, admin overrides without warnings
can manage themes
can add individual scenes to library through share folders (if they have credentials for the smb share)
can reset user tier passwords
everything else forbidden
Admins:
unblocked read write access, can grant/ revoke user and moderator tier permissions
can reset user and moderator passwords
can add new scenes and share folders
Owner account: no restrictions
i guess i could figure out how to fork the main and upload my binary for x64 linux if anyone wanted to toy with it.

**Comment by dillardblom** [comment]
If you would be able to extract a PR from your hard work, I think that would be greatly appreciated! This sounds like already so much more then most of us would be already happy with!
